File: 2024/day09/go/solution.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// parseDiskMap parses the disk map into an expanded block representation.
// Returns a slice where each element is file ID or -1 for free space.
func parseDiskMap(filename string) ([]int, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	diskMap := strings.TrimSpace(string(data))
	blocks := []int{}
	fileID := 0
	isFile := true

	for _, ch := range diskMap {
		length := int(ch - '0')
		if isFile {
			for i := 0; i < length; i++ {
				blocks = append(blocks, fileID)
			}
			fileID++
		} else {
			for i := 0; i < length; i++ {
				blocks = append(blocks, -1) // -1 represents free space
			}
		}
		isFile = !isFile
	}

	return blocks, nil
}

// copyBlocks creates a copy of the blocks slice
func copyBlocks(blocks []int) []int {
	result := make([]int, len(blocks))
	copy(result, blocks)
	return result
}

// compactBlocks compacts disk by moving blocks one at a time from end to leftmost free space
func compactBlocks(blocks []int) []int {
	blocks = copyBlocks(blocks)
	left := 0
	right := len(blocks) - 1

	for left < right {
		// Find leftmost free space
		for left < right && blocks[left] != -1 {
			left++
		}
		// Find rightmost file block
		for left < right && blocks[right] == -1 {
			right--
		}

		if left < right {
			// Swap
			blocks[left] = blocks[right]
			blocks[right] = -1
			left++
			right--
		}
	}

	return blocks
}

// calculateChecksum calculates filesystem checksum: sum of position * file_id for each block
func calculateChecksum(blocks []int) int {
	checksum := 0
	for pos, fileID := range blocks {
		if fileID != -1 {
			checksum += pos * fileID
		}
	}
	return checksum
}

// part1 compacts by moving individual blocks, returns checksum
func part1() (int, error) {
	blocks, err := parseDiskMap("../input.txt")
	if err != nil {
		return 0, err
	}
	compacted := compactBlocks(blocks)
	return calculateChecksum(compacted), nil
}

// part2 compacts by moving whole files (highest ID first), returns checksum
func part2() (int, error) {
	blocks, err := parseDiskMap("../input.txt")
	if err != nil {
		return 0, err
	}

	// Find all files: file_id -> (start_pos, length)
	type FileInfo struct {
		start  int
		length int
	}
	files := make(map[int]FileInfo)
	i := 0
	for i < len(blocks) {
		if blocks[i] != -1 {
			fileID := blocks[i]
			start := i
			for i < len(blocks) && blocks[i] == fileID {
				i++
			}
			files[fileID] = FileInfo{start: start, length: i - start}
		} else {
			i++
		}
	}

	// Find max file ID
	maxFileID := 0
	for fileID := range files {
		if fileID > maxFileID {
			maxFileID = fileID
		}
	}

	// Process files in decreasing order of file ID
	for fileID := maxFileID; fileID >= 0; fileID-- {
		info, exists := files[fileID]
		if !exists {
			continue
		}
		start := info.start
		length := info.length

		// Find leftmost span of free space that fits this file
		// Must be to the left of current position
		freeStart := -1
		i := 0
		for i < start {
			if blocks[i] == -1 {
				// Count consecutive free blocks
				spanStart := i
				spanLength := 0
				for i < start && blocks[i] == -1 {
					spanLength++
					i++
				}
				if spanLength >= length {
					freeStart = spanStart
					break
				}
			} else {
				i++
			}
		}

		// Move file if we found a suitable span
		if freeStart != -1 {
			// Clear old position
			for j := start; j < start+length; j++ {
				blocks[j] = -1
			}
			// Write to new position
			for j := freeStart; j < freeStart+length; j++ {
				blocks[j] = fileID
			}
			// Update file position
			files[fileID] = FileInfo{start: freeStart, length: length}
		}
	}

	return calculateChecksum(blocks), nil
}

func main() {
	result1, err := part1()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error in part1: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 1: %d\n", result1)

	result2, err := part2()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error in part2: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 2: %d\n", result2)
}

File: 2024/day09/c/solution.c
/*
 * Advent of Code 2024 Day 9: Disk Fragmenter
 *
 * Compact a fragmented disk by moving file blocks to fill gaps.
 * Part 1: Move blocks one at a time from end to leftmost free space
 * Part 2: Move whole files to leftmost span that fits
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DISK_MAP 20100
#define MAX_BLOCKS 200000

typedef struct {
    int start;
    int length;
} FileInfo;

/*
 * Parse disk map into expanded block representation.
 * Returns number of blocks created.
 * blocks array where each element is file ID or -1 for free space.
 */
int parse_disk_map(const char *filename, int *blocks) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("Failed to open file");
        exit(1);
    }

    char disk_map[MAX_DISK_MAP];
    if (!fgets(disk_map, sizeof(disk_map), f)) {
        fprintf(stderr, "Failed to read disk map\n");
        exit(1);
    }
    fclose(f);

    // Remove trailing newline
    size_t len = strlen(disk_map);
    if (len > 0 && disk_map[len - 1] == '\n') {
        disk_map[len - 1] = '\0';
        len--;
    }

    int block_count = 0;
    int file_id = 0;
    int is_file = 1;

    for (size_t i = 0; i < len; i++) {
        int digit = disk_map[i] - '0';
        if (digit < 0 || digit > 9) {
            continue; // Skip non-digit characters
        }

        if (is_file) {
            for (int j = 0; j < digit; j++) {
                blocks[block_count++] = file_id;
            }
            file_id++;
        } else {
            for (int j = 0; j < digit; j++) {
                blocks[block_count++] = -1; // -1 represents free space
            }
        }
        is_file = !is_file;
    }

    return block_count;
}

/*
 * Compact disk by moving blocks one at a time from end to leftmost free space.
 */
void compact_blocks(int *blocks, int block_count) {
    int left = 0;
    int right = block_count - 1;

    while (left < right) {
        // Find leftmost free space
        while (left < right && blocks[left] != -1) {
            left++;
        }
        // Find rightmost file block
        while (left < right && blocks[right] == -1) {
            right--;
        }

        if (left < right) {
            // Swap
            blocks[left] = blocks[right];
            blocks[right] = -1;
            left++;
            right--;
        }
    }
}

/*
 * Calculate filesystem checksum: sum of position * file_id for each block.
 */
long long calculate_checksum(const int *blocks, int block_count) {
    long long checksum = 0;
    for (int pos = 0; pos < block_count; pos++) {
        if (blocks[pos] != -1) {
            checksum += (long long)pos * blocks[pos];
        }
    }
    return checksum;
}

/*
 * Part 1: Compact by moving individual blocks, return checksum.
 */
long long part1(const char *filename) {
    int *blocks = malloc(MAX_BLOCKS * sizeof(int));
    if (!blocks) {
        perror("Memory allocation failed");
        exit(1);
    }

    int block_count = parse_disk_map(filename, blocks);
    compact_blocks(blocks, block_count);
    long long checksum = calculate_checksum(blocks, block_count);

    free(blocks);
    return checksum;
}

/*
 * Part 2: Compact by moving whole files (highest ID first), return checksum.
 */
long long part2(const char *filename) {
    int *blocks = malloc(MAX_BLOCKS * sizeof(int));
    if (!blocks) {
        perror("Memory allocation failed");
        exit(1);
    }

    int block_count = parse_disk_map(filename, blocks);

    // Find all files: file_id -> (start_pos, length)
    FileInfo *files = calloc(10000, sizeof(FileInfo));
    if (!files) {
        perror("Memory allocation failed");
        exit(1);
    }

    int max_file_id = -1;
    int i = 0;
    while (i < block_count) {
        if (blocks[i] != -1) {
            int file_id = blocks[i];
            int start = i;
            while (i < block_count && blocks[i] == file_id) {
                i++;
            }
            files[file_id].start = start;
            files[file_id].length = i - start;
            if (file_id > max_file_id) {
                max_file_id = file_id;
            }
        } else {
            i++;
        }
    }

    // Process files in decreasing order of file ID
    for (int file_id = max_file_id; file_id >= 0; file_id--) {
        int start = files[file_id].start;
        int length = files[file_id].length;

        // Find leftmost span of free space that fits this file
        // Must be to the left of current position
        int free_start = -1;
        i = 0;
        while (i < start) {
            if (blocks[i] == -1) {
                // Count consecutive free blocks
                int span_start = i;
                int span_length = 0;
                while (i < start && blocks[i] == -1) {
                    span_length++;
                    i++;
                }
                if (span_length >= length) {
                    free_start = span_start;
                    break;
                }
            } else {
                i++;
            }
        }

        // Move file if we found a suitable span
        if (free_start != -1) {
            // Clear old position
            for (int j = start; j < start + length; j++) {
                blocks[j] = -1;
            }
            // Write to new position
            for (int j = free_start; j < free_start + length; j++) {
                blocks[j] = file_id;
            }
            // Update file position
            files[file_id].start = free_start;
        }
    }

    long long checksum = calculate_checksum(blocks, block_count);

    free(files);
    free(blocks);
    return checksum;
}

int main(void) {
    printf("Part 1: %lld\n", part1("../input.txt"));
    printf("Part 2: %lld\n", part2("../input.txt"));
    return 0;
}
